### PR TITLE
Base of plugin hooks, invoice mutation

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
Invoice request mutation, changes to models, plugin hooks

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [ ] Database queries are optimized and the number of queries is constant.
1. [ ] Database migration files are up to date.
1. [ ] The changes are tested.
1. [ ] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by django-queries** [comment]
Here is the report for c5c84025f126d978d82310d4e29dad224bb61553 (mirumee/saleor @ invoice-request-mutation)
Base comparison is a0b14fb206003f1a7e65d911882cbd74f97c9b5e.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark category
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  category view                              	         23	         23	              0

# api.benchmark checkout
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  add billing address to checkout            	         34	         34	             20
  add shipping to checkout                   	          7	          7	              0
  checkout payment charge                    	         10	         10	              0
  complete checkout                          	          8	          8	              0
  create checkout                            	          5	          5	              1

# api.benchmark collection
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  collection view                            	         21	         21	              0

# api.benchmark homepage
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  featured products list                     	         20	         20	              0
  retrieve main menu                         	          8	          8	              0
  retrieve product list                      	          6	          6	              0
  retrieve secondary menu                    	          8	          8	              0
  retrieve shop                              	          2	          2	              0
  user checkout details                      	         53	         53	             28

# api.benchmark product
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product details                            	         17	         17	              3
  retrieve product attributes                	          9	          9	              0

# api.benchmark variant
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product variant bulk create                	         48	         48	              3
  retrieve variant list                      	         23	         23	              9

# api.benchmark variant stocks
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product variants stocks create             	         19	         19	              4
  product variants stocks delete             	         16	         16	              4
  product variants stocks update             	         23	         23	              4

# api product sorting attributes
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  sort product not having attribute data     	         20	         20	              0
```

</p>
</details>

File: saleor/extensions/manager.py [hunk 1]
```diff
@@ -206,6 +206,12 @@ def order_created(self, order: "Order"):
         default_value = None
         return self.__run_method_on_plugins("order_created", default_value, order)
 
+    def invoice_request(self, order: "Order", invoice: "Invoice", number: str):
```
Comment (korycins) [inline saleor/extensions/manager.py]: `number: optional[str]`

File: saleor/graphql/order/mutations/orders.py [hunk 2]
```diff
@@ -514,3 +515,32 @@ class Meta:
         model = models.Order
         permissions = (OrderPermissions.MANAGE_ORDERS,)
         public = False
+
+
+class RequestInvoice(BaseMutation):
+    class Arguments:
+        order_id = graphene.Int(
+            required=True, description="ID of the order related to invoice."
```
Comment (korycins) [inline saleor/graphql/order/mutations/orders.py]: this is only a detail but required is True by default so no need to add it :)

File: saleor/graphql/order/mutations/orders.py [hunk 3]
```diff
@@ -514,3 +515,32 @@ class Meta:
...
+        except models.Order.DoesNotExist:
+            raise ValidationError(
+                "Order with provided id does not exist.", code=OrderErrorCode.NOT_FOUND,
+            )
+
+        invoice = models.Invoice.objects.create(status=InvoiceStatus.PENDING)
+        info.context.extensions.invoice_request(
+            order=order, invoice=invoice, number=data.get("number")
+        )
+        return RequestInvoice()
```
Comment (korycins) [inline saleor/graphql/order/mutations/orders.py]: Don't we want to return something here?

Comment (tomaszszymanski129) [reply saleor/graphql/order/mutations/orders.py]: What would you suggest? after all this is a request of the invoice creation where the process starts. It has nothing to return, it either throws an error or it doesn't.

File: saleor/graphql/order/mutations/orders.py (hunk 2 above)
Comment (tomaszszymanski129) [reply saleor/graphql/order/mutations/orders.py]: is it? if i remove it and re-generate the schema it's missing ! mark near this field

Comment (maarcingebala) [reply saleor/graphql/order/mutations/orders.py]: @korycins Default is False ;)

Comment (maarcingebala) [reply saleor/graphql/order/mutations/orders.py]: Whatever it is, I find it useful to have it explicitly defined.

File: saleor/graphql/order/mutations/orders.py [hunk 4]
```diff
@@ -514,3 +515,32 @@ class Meta:
         model = models.Order
         permissions = (OrderPermissions.MANAGE_ORDERS,)
         public = False
+
+
+class RequestInvoice(BaseMutation):
+    class Arguments:
+        order_id = graphene.Int(
```
Comment (maarcingebala) [inline saleor/graphql/order/mutations/orders.py]: We rather want to use `graphene.ID` here.

File: saleor/graphql/order/mutations/orders.py (hunk 3 above)
Comment (maarcingebala) [reply saleor/graphql/order/mutations/orders.py]: Don't we want to use Job objects here? The idea behind them is to represent async actions like generating files etc.

Comment (maarcingebala) [reply saleor/graphql/order/mutations/orders.py]: It can also be added as a separate PR.

**Comment by maarcingebala** [comment]
@tomaszszymanski129 I suggest that we create a feature branch first and we'll merge this and future PRs related to invoices, to keep `master` unaffected. Once the entire feature is ready we'll ship it to master.

**Comment by tomaszszymanski129** [comment]
@maarcingebala Changed the mutation field to Graphene ID and removed Django ORM code in favor of Graphene style node fetching. Changed merge branch to feature branch instead of master. Plan is to add the Job class in one of next tickets, if this looks good can be merged to feature branch.

**Pull request merged**